src/Microsoft.Data.SqlClient/add-ons/AzureKeyVaultProvider/AzureSqlKeyCryptographer.cs

```diff
@@ -0,0 +1,232 @@
+            cryptoClientDictionary[keyIdentifierUri] = cryptographyClient;
+
+            return cryptographyClient;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="vaultUri">The Azure Key Vault URI</param>
+        /// <param name="keyName">The name of the Azure Key Vault key</param>
+        /// <param name="keyResourceUri">The Azure Key Vault key identifier</param>
+        private void FetchKey(Uri vaultUri, string keyName, string keyResourceUri)
+        {
+            var fetchKeyTask = FetchKeyFromKeyVault(vaultUri, keyName);
+            _keyFetchTaskDictionary[keyResourceUri] = fetchKeyTask;
+
+            fetchKeyTask
+                .ContinueWith(k => ValidateRsaKey(k.Result))
+                .ContinueWith(k => _keyDictionary[keyResourceUri] = k.Result);
+
+            Task.Run(() => fetchKeyTask);
+        }
+
+        /// <summary>
+        /// Looks up the KeyClient object by it's URI and then fethces the key by name.
+        /// </summary>
+        /// <param name="vaultUri">The Azure Key Vault URI</param>
+        /// <param name="keyName">Then name of the key</param>
+        /// <returns></returns>
+        private Task<Azure.Response<KeyVaultKey>> FetchKeyFromKeyVault(Uri vaultUri, string keyName) => keyClientDictionary[vaultUri].GetKeyAsync(keyName);
+
+        /// <summary>
+        /// Validates that a key is of type RSA
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns></returns>
+        private KeyVaultKey ValidateRsaKey(KeyVaultKey key)
+        {
+            if (key.KeyType != KeyType.Rsa && key.KeyType != KeyType.RsaHsm)
+            {
+                throw new FormatException(string.Format(CultureInfo.InvariantCulture, Strings.NonRsaKeyTemplate, key.KeyType));
+            }
+
+            return key;
+        }
+
+        /// <summary>
+        /// Instantiates and adds a KeyClient to the KeyClient dictionary
+        /// </summary>
+        /// <param name="vaultUri">The Azure Key Vault URI</param>
+        private void CreateKeyClient(Uri vaultUri)
+        {
+            if (!keyClientDictionary.ContainsKey(vaultUri))
+            {
+                keyClientDictionary[vaultUri] = new KeyClient(vaultUri, TokenCredential);
+            }
+        }
+
+        /// <summary>
+        /// Validates and parses the Azure Key Vault URI and key name.
+        /// </summary>
+        /// <param name="masterKeyPath">The Azure Key Vault key identifier</param>
+        /// <param name="vaultUri">The Azure Key Vault URI</param>
+        /// <param name="masterKeyName">The name of the key</param>
+        private void ParseAKVPath(string masterKeyPath, out Uri vaultUri, out string masterKeyName)
+        {
+            Uri masterKeyPathUri = new Uri(masterKeyPath);
+            vaultUri = new Uri(masterKeyPathUri.GetLeftPart(UriPartial.Authority));
+            masterKeyName = masterKeyPathUri.Segments[2];
```

Comment:
Is the master key path getting validated that it is the right format somewhere? We are not sure that Segment[2] exists by the time we are here, right? I think we were validating these things when the key URI was entering via the public provider interface methods but I do not see where the validation is happening.

Update: I later found that this was in the `ValidateNonEmptyAKVPath`